Add parallel interface (DVP) support for ov5640 driver

CharlesDias opened the pull request:
Related to the issue https://github.com/zephyrproject-rtos/zephyr/issues/74353.

These PRs are requires:
- https://github.com/zephyrproject-rtos/zephyr/pull/72254
- https://github.com/zephyrproject-rtos/zephyr/pull/74415.
- https://github.com/zephyrproject-rtos/zephyr/pull/76144.
- https://github.com/zephyrproject-rtos/zephyr/pull/71463

This PR is related to:
-  https://github.com/zephyrproject-rtos/zephyr/pull/76143

Testing on STM32H7B3I-DK

~~`west build -p -b stm32h7b3i_dk --shield st_b_cams_omv_mb1683 samples/subsys/video/capture_to_lvgl/`~~
`west build -p -b stm32h7b3i_dk --shield st_b_cams_omv_mb1683 samples/drivers/video/capture_to_lvgl/`

https://github.com/zephyrproject-rtos/zephyr/assets/27859307/1bf60853-ae4d-4bd8-a28b-a6dcab5d5a8f

CharlesDias commented:
Dear,

Please, @ngphibang, @josuah, @erwango, and @loicpoulain, could you take a look at the draft PR to check if I am on the right path?

Thanks!

josuah commented on drivers/video/ov5640.c:
```diff
@@ -457,23 +635,84 @@ static int ov5640_write_multi_regs(const struct i2c_dt_spec *spec, const struct
+
+	if (ret) {
+		LOG_ERR("Unable to configure REG: %d on DVP", TIMING_TC_REG21_REG);
+		return ret;
```
If `read()` and `write()` have different errors, then `ret` will not contain a valid error code, so cannot be returned directly... I am still unsure of the best way to handle this while avoiding one error message per line. Maybe this?

```
if ((ret = ov5640_read_reg(&cfg->i2c, TIMING_TC_REG21_REG, &reg, sizeof(reg))) ||
    (ret = ov5640_write_reg(&cfg->i2c, TIMING_TC_REG21_REG, (reg & 0xDF) | 0x00))) {
        LOG_ERR(...);
        return ret;
}
```

Although this style is also rare in Zephyr.

In this particular case, though, this sounds like solved by `ov5640_modify_reg()`.

josuah replied:
As seen below, you can also use a const table (i.e. global at the top) of addresses/value/mask to apply.

CharlesDias replied:
> In this particular case, though, this sounds like solved by `ov5640_modify_reg()`.

`ov5640_modify_reg()` sounds great! I'll change to use it.

josuah replied:
Looking again at various image sensor drivers, the `ret |= ...` seems present in various sensor.
Contributions tend to preserve it, but do feel free to break these local changes for the more Zephyr-like `ret = ...; if (ret) { return ret; }` for new code. :)

josuah commented on drivers/video/ov5640.c:
```diff
@@ -457,23 +635,84 @@ static int ov5640_write_multi_regs(const struct i2c_dt_spec *spec, const struct
+	int ret = 0;
+
+	/* Configure MIPI pixel clock */
+	ret |= ov5640_modify_reg(&cfg->i2c, SC_PLL_CTRL0_REG, 0x0f, 0x08);
+	ret |= ov5640_modify_reg(&cfg->i2c, SC_PLL_CTRL1_REG, 0xff,
+				 resolution_params->mipi_pclk.pllCtrl1);
+	ret |= ov5640_modify_reg(&cfg->i2c, SC_PLL_CTRL2_REG, 0xff,
+				 resolution_params->mipi_pclk.pllCtrl2);
+	ret |= ov5640_modify_reg(&cfg->i2c, SC_PLL_CTRL3_REG, 0x1f, 0x13);
+	ret |= ov5640_modify_reg(&cfg->i2c, SYS_ROOT_DIV_REG, 0x3f, 0x01);
+	ret |= ov5640_write_reg(&cfg->i2c, PCLK_PERIOD_REG, 0x0a);
```
Minor style issue, like previous comment, but for this time, can be handled by a local table of register/value to apply, using `0xff` for the last one eventually, wasting 1 I2C transaction for when a write can immediately be made.

CharlesDias replied:
This code snippet was modified after rebase with https://github.com/zephyrproject-rtos/zephyr/pull/76144.

josuah replied:
Yes my bad, I identified the one commit this PR proposes. Marking resolved. Thanks!

josuah commented on drivers/video/ov5640.c:
```diff
@@ -457,23 +635,84 @@ static int ov5640_write_multi_regs(const struct i2c_dt_spec *spec, const struct
 {
 	struct ov5640_data *drv_data = dev->data;
 	const struct ov5640_config *cfg = dev->config;
+	const struct video_format_cap *fmts = ov5640_is_dvp(dev) ? dvp_fmts : csi2_fmts;
+	const size_t num_fmts = ov5640_is_dvp(dev) ? ARRAY_SIZE(dvp_fmts) : ARRAY_SIZE(csi2_fmts);
+	const struct ov5640_resolution_config *resolution_params = ov5640_is_dvp(dev) ?
+						dvp_resolution_params : csi2_resolution_params;
+	const size_t num_res_params = ov5640_is_dvp(dev) ?
+			ARRAY_SIZE(dvp_resolution_params) : ARRAY_SIZE(csi2_resolution_params);
```
Minor style comment: I am tempted by a single `if/else` or `switch (cfg->bus_type)` to help with readability and error handling.

CharlesDias replied:
Yeap. It's better. I'll change to `if/else`.

josuah commented on drivers/video/ov5640.c:
```diff
@@ -644,19 +890,33 @@ static int ov5640_init(const struct device *dev)
+	}
+
 	if (ret) {
-		LOG_ERR("Unable to set virtual channel");
+		LOG_ERR("Unable to initialize the sensor with DVP parameters");
 		return -EIO;
 	}
```
This looks all correct, but it would be less confusing to have it inside the if to group it semantically.

CharlesDias replied:
Done.

josuah commented on drivers/video/ov5640.c:
```diff
@@ -325,6 +334,114 @@ static const struct ov5640_reg ov5640InitParams[] = {
 	{0x5000, 0xa7},
 };
 
+/* FIXME: Check with values are really necessary. */
```
Just my impression: Hard to know for each of them, and also true for the other parameters above, so might be possible to remove that comment, with more pull requests to characterize more of these registers.

CharlesDias replied:
Done.

josuah commented on drivers/video/ov5640.c:
```diff
@@ -669,10 +929,18 @@ static int ov5640_init(const struct device *dev)
-	fmt.width = 1280;
-	fmt.height = 720;
+	if (ov5640_is_dvp(dev)) {
+		/* Set default format to 480x272 RGB565 */
```
Comment: This looks like one particular display rather than a common resolution format, though it is also good that the defaults did get tested on actual hardware.

CharlesDias replied:
Since this version of the DVP feature does not support the current resolution configuration (1280x720), it would be better to set it up for the minimal resolution supported by DVP.

josuah replied:
Ah right I understand better now thank you.
Did you try various resolutions and noticed that 480x272 RGB565 was the highest one supported by DVP?
I do not see this resolution anywhere on [Linux](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/media/i2c/ov5640.c) or the [datasheet](https://cdn.sparkfun.com/datasheets/Sensors/LightImaging/OV5640_datasheet.pdf).
Just curious. :)

CharlesDias replied:
> Ah right I understand better now thank you.
> Did you try various resolutions and noticed that 480x272 RGB565 was the highest one supported by DVP?
> I do not see this resolution anywhere on [Linux](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/media/i2c/ov5640.c) or the [datasheet](https://cdn.sparkfun.com/datasheets/Sensors/LightImaging/OV5640_datasheet.pdf).
> Just curious. :)

DVP should support resolutions higher than 480x727, but due to my board's memory (I'm using only the internal SRAM) and display limitations, I was able to test 160x120, 320x240, and 480x272.

For future version, I'll try to use the external SDRAM and others resolution.

josuah commented on drivers/video/ov5640.c:
```diff
@@ -325,6 +334,114 @@ static const struct ov5640_reg ov5640InitParams[] = {
+	{0x3a14, 0x07},
+	{0x3a15, 0xae},
+
+	/* JPEG Control */
+	{0x4401, 0x0d}, /* | Read SRAM enable when blanking | Read SRAM at first blanking */
+	{0x4723, 0x03}, /* DVP JPEG Mode456 Skip Line Number */
```
Did you try without these? The sensor supports outputting JPEG directly, but you did not enable it via `0x3821 COMPRESSION ENABLE Bit[5]: JPEG enable`.

If the image still works without it, this might be possible to remove these to reduce confusion.

Though, it is also possible that it is not *only* related to JPEG...

josuah replied:
https://github.com/zephyrproject-rtos/zephyr/pull/76124/files#r1685824413

CharlesDias replied:
> Did you try without these? The sensor supports outputting JPEG directly, but you did not enable it via `0x3821 COMPRESSION ENABLE Bit[5]: JPEG enable`.
> 
> If the image still works without it, this might be possible to remove these to reduce confusion.
> 
> Though, it is also possible that it is not only related to JPEG...

I removed it and it stayed the same.

josuah commented on drivers/video/ov5640.c:
```diff
@@ -325,6 +334,114 @@ static const struct ov5640_reg ov5640InitParams[] = {
+	{0x3c0a, 0x9c},
+	{0x3c0b, 0x40},
+	{TIMING_TC_REG20_REG, 0x47},
+	{TIMING_TC_REG21_REG, 0x01},
```
Note: this disables JPEG

josuah reviewed (commented):
Thank you for this addition!
Progressively, more features of these complex sensors can be integrated from Linux and other sources into Zephyr.

This looks like almost ready to go. See some few feedback inline. Feel free to dismiss it or address it as relevant.

josuah commented:
This might need rebasing once https://github.com/zephyrproject-rtos/zephyr/pull/76144 (not in draft state) is merged.
Though, this is currently feature-freeze for `v3.7.0`

ngphibang commented on dts/bindings/video/ovti,ov5640.yaml:
```diff
@@ -5,7 +5,7 @@ description: OV5640 CMOS video sensor
 compatible: "ovti,ov5640"
 
-include: i2c-device.yaml
+include: [i2c-device.yaml, video-interfaces.yaml]
```
I think this will not have any effects as the properties defined in `video-interfaces.yaml` are for an**endpoint** node, not for the ov5640 node itself. The difficulty here (and the reason why I let https://github.com/zephyrproject-rtos/zephyr/pull/74415 always as draft) is that we cannot know the **endpoint** is actually a child or a grand child node of the video device node (e.g. ov5640) as **ports** node is optional, so that we don't know using 1 or 2 levels of **child-binding** inb the video-interfaces.yaml:

- child-binding:
     child-binding:

So, I think the solution here is that either we:
- Don't include` video-interfaces.yaml`, and just declare the properties in the devicetree. It will compile without any problem.
- Or add `compatible: "nxp,zephyr-endpoint"` in the endpoint node and in `video-interfaces.yaml`

@josuah Do you have a solution for this ?

CharlesDias replied:
Dear, @ngphibang and @josuah,

If I change from `include: [i2c-device.yaml, video-interfaces.yaml] `to `include: i2c-device.yaml`, I got the compile error below:


`devicetree error: 'bus-type' appears in /soc/i2c@58001c00/ov5640@3c in /home/charlesdias/zephyrproject/zephyr/build/zephyr/zephyr.dts.pre, but is not declared in 'properties:' in /home/charlesdias/zephyrproject/zephyr/dts/bindings/video/ovti,ov5640.yaml`

ngphibang replied:
I think it's because you declared these properties on the ov5640 node. All the properties in `video-interfaces.yaml` should be declared in the **endpoint** subnode.

CharlesDias commented:
Thanks, @josuah and @ngphibang for your suggestions. I'll fix them during the weekend.

ngphibang commented:
@CharlesDias Thanks for adding this dvp mode to the ov5640 driver. As being said last time in your raised issue, we were also working on supporting changing frame rates and some controls to the ov5640 driver at the same time. Here is [the PR](https://github.com/zephyrproject-rtos/zephyr/pull/76144). Sorry that we couldn't push our PR earlier to facilitate your work due to some difficulties in the impelmentation.

If possible, could you take into account our changes ?  (IMO, some commits in our PR could make the dvp support easier and your PR is a bigger change so it seems easier to rebase your change on ours than vice versa ?)

I am available for any helps if needed.

CharlesDias commented:
> If possible, could you take into account our changes ? (IMO, some commits in our PR could make the dvp support easier and your PR is a bigger change so it seems easier to rebase your change on ours than vice versa ?)

For sure, @ngphibang, I'll do the rebase by the end of the week. Thanks for being available!

CharlesDias commented:
Rebase with https://github.com/zephyrproject-rtos/zephyr/pull/76144

CharlesDias commented:
Apply review suggestions.

josuah reviewed (commented):
Thank you for this other cycle of modifications!

One remark as I learn the new `.bus_type` API myself.
If you wish it is possible to convert the PR to a draft and remove `[DNM]` to say that this is a work in progress not to be merged yet.
![2024-09-15-211231_479x349_scrot](https://github.com/user-attachments/assets/3d5c502c-5cb8-42ee-be08-ddf047b6324b)

CharlesDias commented:
> If you wish it is possible to convert the PR to a draft and remove `[DNM]` to say that this is a work in progress not to be merged

Hi @josuah, I considered this initial version ready to go (apart from the reviewer's suggestions and the PR dependencies). I added the [DNM] to avoid merging it before the dependencies. I'm not sure if this is the right approach.

josuah commented:
> initial version ready [...] avoid merging it before the dependencies [...] the right approach

The DNM label gives me a hint that this might be made for this purpose but I would not bet an arm on it...
Learning a bit every day!

CharlesDias committed:
drivers: video: ov5640: add DVP support

Improve the ov5640 video driver to provide parallel interface (DVP) support

Signed-off-by: Charles Dias <charlesdias.cd@outlook.com>

CharlesDias commented:
Hi @josuah and @ngphibang. 

Since PR https://github.com/zephyrproject-rtos/zephyr/pull/76144 has been merged, I'm returning to work on this PR. Could you please explain how to retrieve the driver's bus-type? Considering that I have the DT file below:

Thanks!

```
&i2c4 {
	pinctrl-0 = <&i2c4_scl_pd12 &i2c4_sda_pd13>;
	pinctrl-names = "default";
	status = "okay";
	clock-frequency = <I2C_BITRATE_FAST>;

	ov5640: ov5640@3c {
		compatible = "ovti,ov5640";
		reg = <0x3c>;
		status = "okay";
		reset-gpios = <&gpioa 0 GPIO_ACTIVE_LOW>;
		powerdown-gpios = <&gpioa 7 GPIO_ACTIVE_HIGH>;

		port {
			ov5640_ep_out: endpoint {
				compatible = "zephyr,video-interfaces";
				remote-endpoint-label = "dcmi_ep_in";
				bus-type = "parallel";
			};
		};
	};
};

&dcmi {
	status = "okay";
	pinctrl-0 = <&dcmi_hsync_pa4 &dcmi_pixclk_pa6 &dcmi_vsync_pb7
		&dcmi_d0_pc6 &dcmi_d1_pc7 &dcmi_d2_pg10 &dcmi_d3_pc9
		&dcmi_d4_pc11 &dcmi_d5_pd3 &dcmi_d6_pb8 &dcmi_d7_pb9>;
	pinctrl-names = "default";

	sensor = <&ov5640>;
	bus-width = <8>;
	hsync-active = <0>;
	vsync-active = <0>;
	pixelclk-active = <1>;
	capture-rate = <1>;

	port {
		dcmi_ep_in: endpoint {
			compatible = "zephyr,video-interfaces";
			remote-endpoint-label = "ov5640_ep_out";
		};
	};

	dmas = <&dma1 0 75 (STM32_DMA_PERIPH_TO_MEMORY | STM32_DMA_PERIPH_NO_INC |
		STM32_DMA_MEM_INC | STM32_DMA_PERIPH_8BITS | STM32_DMA_MEM_32BITS |
		STM32_DMA_PRIORITY_HIGH) STM32_DMA_FIFO_1_4>;
};

```

ngphibang commented:
Hi @CharlesDias ,

Yes, sure. Since there is a small change in the [dt-bindings PR](https://github.com/zephyrproject-rtos/zephyr/pull/74415) according to the discussion with the reviewers, to use the new bindings, we need to include the .yaml file instead of using compatible as before. You can find an example of usage here:

https://github.com/zephyrproject-rtos/zephyr/pull/76475/commits/0b29445d072be8aca2e3b28a56eaafd8695034f2

Thank you !

CharlesDias commented:
Thanks, @ngphibang for your help! Could you test this current version on your board? I hope that I haven't broken anything! :) 

In this current version, I just add the basic support for DVP interface. Some improvements related to function control may be made in future PRs.

ngphibang commented:
@CharlesDias It works fine on our side. I need to look through the code though. Will come back soon ...

In the future, do you consider adding support for JPEG format together with some other controls you planed :-) ? I think it's a great feature as it allows to send video frames over network, for example. We 'd like to do that but unfortunately our receiver (mipicsi2rx) does not support JPEG format.

CharlesDias commented:
> In the future, do you consider adding support for JPEG format together with some other controls you planed :-) ?

Hi @ngphibang. I hadn't thought about JPEG yet, but I think I'll be able to look into it in December.

ngphibang commented on drivers/video/ov5640.c:
```diff
@@ -897,15 +1127,15 @@ static int ov5640_enum_frmival(const struct device *dev, enum video_endpoint_id
 {
```
`ov5640_enum_frmival()` is common for both modes, so we cannot return the results of CSI2 for DVP. You need to implement `enum_frmival() ` for DVP by returning the supported framerates for each supported format (I suppose DVP only supports 30 fps for all formats as of now ?).

- The same for `get_frmival()`, it needs to return the current frame rate in DVP modes
- For `set_frmival()`, we can return -ENOTSUP

josuah replied:
Maybe branching on `ov5640_is_dvp()` like you did elsewhere would work well.

CharlesDias replied:
@ngphibang and @josuah

As you commented, it looks like that the 30FPs is used as default for DVP. In DVP the framerate depends on pixel clock. Currently, the init_params_dvp configured the `SC_PLL_CTRL2_REG` with 0x64 (PLL multiplier 100) and `SC_PLL_CTRL3_REG` with 0x13 (PLL root divided by 2 and PLL pre-divider 3). I'll try to measure the current FPS.

The functions `set_frmival, get_frmival`, and `enum_frmival` haven't been implemented on `video_stm32_dcmi.c` driver. That is the reason for those functions haven't broken when the DVP is used. [Check here.](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/video/video_stm32_dcmi.c#L375)

So, since those functions are not implemented on DCMI driver, is this a block to accept this PR? Could this be done in another PR? Thanks!

ngphibang replied:
Thank you for the update. No, it's not a blocker. But as being said in my previous comment, we don't need to implement those functions but is it possible to just return a (hardcode) 30 fps for `enum_frmival()` and `get_frmival()` and -ENOTSUP for `set_frmival()` ?

ngphibang replied:
Well, I've just rethought. Let the frmival callback not implemented is OK as well (especially now we have the new DEVICE_API, it's safe to do so I think). BTW, should the frmival callback in the ov5640 driver be removed as well if it is not tested (due to frmival in dcmi not implemented) ?

CharlesDias replied:
> Well, I've just rethought. Let the frmival callback not implemented is OK as well (especially now we have the new DEVICE_API, it's safe to do so I think). BTW, should the frmival callback in the ov5640 driver be removed as well if it is not tested (due to frmival in dcmi not implemented) ?

Hi @ngphibang.

Regarding the DCMI, when those functions are calling an error is returned, as expected. Considering those frmival functions are working with CSI2 driver, I am not sure If it is necessary to remove. 

```
[00:00:00.168,000] <err> main: Unable to get video frame interval
[00:00:00.168,000] <err> main: Unable to enumerate video frame interval
[00:00:00.168,000] <err> main: Unable to set video frame interval
```

josuah replied:
How about returning `-ENOTSUP` for DCMI in the meantime they get implemented (compromise)?

CharlesDias replied:
@josuah, 

Those functions are not implemented in DCMI driver (`video_stm32_dcmi.c `, [see here](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/video/video_stm32_dcmi.c#L375)). In this way, when those functions are called, the VIDEO API returns `-ENOSYS`. So, to return `-ENOTSUP` I'll need to change the DCMI driver to add some functions that are not supported yet. I'm not sure about.

ngphibang commented on drivers/video/ov5640.c:
```diff
@@ -980,6 +1210,13 @@ static int ov5640_init(const struct device *dev)
 
 	k_sleep(K_MSEC(20));
 
+	/* Reset all registers */
```
What is this ? We need it for both modes ?

CharlesDias replied:
This returns the registers to default values. Should I remove it?

josuah replied:
No idea if it was required for DVP, but this can help to have it for CSI too if the MCU is rebooted but the image sensor is still powered on, such as during firmware upgrades.

josuah replied:
This could help with both CSI and DVP if the MCU/MPU is rebooted but the image sensor is still powered, i.e. brown-out or firmware upgrade.

CharlesDias replied:
@ngphibang, what you think?

ngphibang replied:
yes

ngphibang commented on drivers/video/ov5640.c:
```diff
@@ -650,6 +875,11 @@ static int ov5640_stream_start(const struct device *dev)
 static int ov5640_stream_stop(const struct device *dev)
 {
```
For DVP mode, I think we need to implement stream _start / stop_ also. Pipeline should be capable to stop and restart. Could you refer to the Linux implementation [here](https://elixir.bootlin.com/linux/v6.11.1/source/drivers/media/i2c/ov5640.c#L1821)

CharlesDias replied:
Sorry, @ngphibang. I've forgotten to comment that this is done.

CharlesDias commented:
Removed `zephyr/dts/bindings/video/ovti,ov5640.yaml` changes.

CharlesDias commented on dts/bindings/video/ovti,ov5640.yaml:
> * This change should be in a separate commit that says something like "ov5640: Switch to the new video interfaces binding", as it does not reveal exactly why we need this for dvp support and the scope of the change is also larger than just for DVP support.

Removed this change from `ovti,ov5640.yaml`

CharlesDias commented:
Rebase and fixed conflict.

CharlesDias commented on drivers/video/ov5640.c:
```diff
@@ -1017,8 +1268,15 @@ static int ov5640_init(const struct device *dev)
-	fmt.width = 1280;
-	fmt.height = 720;
+	if (ov5640_is_dvp(dev)) {
+		/* Set default format to QQVGA (160x120) RGB565 */
```
The DVP does not support the 720p resolution so far.

CharlesDias replied:
Fixed. Thanks!

CharlesDias commented on drivers/video/ov5640.c:
```diff
@@ -1017,8 +1268,15 @@ static int ov5640_init(const struct device *dev)
+		fmt.width = 160;
+		fmt.height = 120;
+	} else {
+		/* Set default format to 720p RGB565 */
```
Fixed.

CharlesDias commented on drivers/video/ov5640.c:
Change to csi2.

josuah commented on drivers/video/ov5640.c:
```diff
@@ -989,17 +1225,31 @@ static int ov5640_init(const struct device *dev)
+
 	/* Set virtual channel */
-	ret = ov5640_modify_reg(&cfg->i2c, 0x4814, 3U << 6, (uint8_t)(DEFAULT_MIPI_CHANNEL) << 6);
-	if (ret) {
-		LOG_ERR("Unable to set virtual channel");
-		return -EIO;
+	if (!ov5640_is_dvp(dev)) {
+		ret = ov5640_modify_reg(&cfg->i2c, 0x4814, 3U << 6,
+					(uint8_t)(DEFAULT_MIPI_CHANNEL) << 6);
```
Absolutely non-blocking, minor optimization: create a `init_params_csi2` table where all MIPI/CSI2 parameters can be added.
```suggestion
} else {
		ret = ov5640_write_multi_regs(&cfg->i2c, init_params_csi2,
					      ARRAY_SIZE(init_params_csi2));
		if (ret) {
			LOG_ERR("Unable to set virtual channel");
			return -EIO;
		}
```

This is functionally equivalent to what you did though!

CharlesDias replied:
The MIPI/CSI2 parameters are already into the `init_params_common`. Maybe in the future could be interesting to check those parameters carefully to reduce unnecessary parameter configuration, as you mentioned below.

josuah commented on drivers/video/ov5640.c:
```diff
@@ -359,14 +362,117 @@ static const struct ov5640_reg init_params[] = {
+	{0x3c0b, 0x40},
+	{TIMING_TC_REG20_REG, 0x47},
+	{TIMING_TC_REG21_REG, 0x01},
+	{0x3800, 0x00},
```
If I understand correctly, the way to initialize DVP is to first initialize MIPI/CSi2, then change registers to become DVP.

This is not a problem though. If someone bother the small extra ROM over head or initialization time they can then send a patch :)

josuah reviewed (changes requested):
Some minor remarks that you can ignore as you wish.
The only bug is to have CSI2 framerates returned instead of DVP, though, apparently, that does not break DVP mode.

Thank you for extending the OV5640 support beyond just MIPI! Zephyr is full of small devices and MIPI-capable devices can often run Linux.

josuah commented on tests/drivers/build_all/video/app.overlay:
```diff
@@ -53,6 +53,13 @@
+				port {
+					ov5640_ep_out: endpoint {
+						remote-endpoint-label = "mipi_csi2rx_ep_in";
+						bus-type = <4>;
```
Using [`VIDEO_BUS_TYPE_CSI2_DPHY`](https://github.com/zephyrproject-rtos/zephyr/blob/baa49f6f32d5632c50c2c65dda44493dfe8ef8a3/include/zephyr/dt -bindings/video/video-interfaces.h#L13) could look a little bit more friendly. :) Tests can act as documentation too.

CharlesDias replied:
Done.

CharlesDias commented:
Implemented partial @ngphibang and @josuah's suggestions. I still need to work on `ov5640_enum_frmival()`, `get_frmival()`, and `set_frmival()` implementation. Those function updates are missing!

josuah commented on drivers/video/ov5640.c:
```diff
@@ -131,6 +133,7 @@ struct ov5640_mipi_frmrate_config {
 struct ov5640_mode_config {
 	uint16_t width;
 	uint16_t height;
+	uint16_t array_size_res_params;
```
While seen in context (`ARRAY_SIZE(res_params)`) this makes sense, but here, `res_params_size` or `_len` or `_num` would be more intuitive. Not blocking at all!

josuah commented on tests/drivers/build_all/video/mimxrt1064_evk.overlay:
This seems accidentally added, could you confirm it is still needed?

josuah commented on tests/drivers/build_all/video/mimxrt1064_evk.overlay:
```diff
@@ -0,0 +1,51 @@
+
+				port {
+					ov5640_ep_out: endpoint {
+						remote-endpoint-label = "mipi_csi2rx_ep_in";
```
Maybe having a video RX peripheral remote-endpoint was the reason why you needed this file.
There is an emulated/fake video RX peripheral which you might be able to use for this purpose if you wish to remain hardware independent.

I do not see how using NXP hardware would be an issue for the build-all test though.

ngphibang replied:
@CharlesDias @josuah

- Why do you add NXP imxrt1064 overlay to test ov5640 dvp mode ? Is this HW support ov5640 camera sensor ? AFAIU, this should be an ST board (with DCMI) ?
- This is a built test so I think we need to make a **real board overlay** to involve the actual related drivers and actual code into the test (for imxrt1170 : ov5640 in csi2 mode --> mipicsi2rx --> csi, for ST board : ov5640 in dvp --> DCMI) as the purpose of a test is to find bugs in real drivers. Another reason not to use the emulated video drivers.

josuah replied:
> could you please share more information about this "emulated/fake video RX peripheral"? Thanks!

I missed that message sorry. Here is the PR that introduced it for your own curiosity
https://github.com/zephyrproject-rtos/zephyr/pull/79482

Here is an example of integration for testing the  API itself:
https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/video/api

This PR got merged a bit fast: I am not against replacing it if something else (i.e. sw_generator) makes more sense!

> it will test either csi2 or dvp mode
> this should be an ST board (with DCMI)

Not sure why I did not think of it, DCMI as a target sounds ideal.
I will be a bit busy for the next 3 weeks, so cannot offer the resources for it until then.

josuah reviewed (changes requested):
I think these are the two unresolved points that could need action. The rest seems ready to land AFAICT.

https://github.com/zephyrproject-rtos/zephyr/pull/76124/files#r1806182604
https://github.com/zephyrproject-rtos/zephyr/pull/76124/files#r1806168527

Thanks once again.

CharlesDias commented:
> I think these are the two unresolved points that could need action. The rest seems ready to land AFAICT.
> 
> https://github.com/zephyrproject-rtos/zephyr/pull/76124/files#r1806182604
> https://github.com/zephyrproject-rtos/zephyr/pull/76124/files#r1806168527

Hi, @josuah,

The first one is done. I leave some comments on the second. Thanks for your review!

CharlesDias commented:
> Oops, sorry, please forget my above nonsense comment, it seems coming from an AI-chatbot not me (maybe I was in hurry while writting this :( ).
> 
> My point was just that you can leave the framerate implementation for the future, it's not a blocking point. But since ov5640 driver now supports both dvp and csi2 modes, leaving it as-is will make people think that set/get/enum_frival() is actually working / valid for both modes while it is not the case here.

@ngphibang, I added the
```
if (ov5640_is_dvp()) {
    return -ENOTSUP;
}
```
inside of `set/get/enum_frival()` to be clear that these functions do not support the DVP. I also added inside the get because I am not sure that the returned value is correct.

CharlesDias commented on tests/drivers/build_all/video/app.overlay:
```diff
@@ -53,6 +55,13 @@
 				reset-gpios = <&test_gpio 0 0>;
 				powerdown-gpios = <&test_gpio 1 0>;
+
+				port {
```
@ngphibang, thanks! This test framework is a mysterious for me! :) Do you have any reference material when I can learn more about it?

> BTW to use the port and endpoint, you need to modify the DCMI driver to use the new video-interface bindings (it's already done in another PR if I remember well ?)

Correctly. This was made by another PR. I haven't had any need to modify the DCMI driver so far.

CharlesDias replied:
@ngphibang and @josuah, I removed the imxrt1064 overlay and created on for stm32h7b3i_dk. However, two twister test failed. The [twister-build (2)](https://github.com/zephyrproject-rtos/zephyr/actions/runs/12966433889/job/36167274237?pr=76124#logs) complained against imxrt1064 and [twister-build (3)](https://github.com/zephyrproject-rtos/zephyr/actions/runs/12966433889/job/36167274319?pr=76124#logs) against native_sim. 

Could you please give me some help? Thanks!

ngphibang replied:
@CharlesDias : It's because in the ov5640 driver, you get the bus type with:  

`.bus_type = DT_PROP(DT_CHILD(DT_INST_CHILD(n, port), endpoint), bus_type),`

but when the ov5640 driver is tested on the native_sim (drivers.video.build test) and the mimxrt1064_evk (drivers.video.mcux_csi.build test) platforms, the property `bus-type` is not defined.

So, to fix it, I think the ov5640 node should be removed in the app.overlay as ov5640 driver tests are already covered in drivers.video.mcux_csi.build, drivers.video.mcux_csi2rx.build and drivers.video.stm32_dcmi.build (the one you will add) tests

CharlesDias replied:
The bus_type is initialized with the default value VIDEO_BUS_TYPE_CSI2_DPHY when bus-type property is not declared in the device tree. This way, I don't need to make changes on the video test folder for now.

I'll add the test case for DCMI driver in another PR. I have to study the test framework fir

ngphibang commented on tests/drivers/build_all/video/stm32h7b3i_dk.overlay:
```diff
@@ -0,0 +1,49 @@
+
+				port {
+					ov5640_ep_out: endpoint {
+						remote-endpoint-label = "dvp_ep_in";
```
Where is the `dvp_ep_in` ? I think you need to add the receiver (DCMI) node. Moreover, a test case should be added for the dcmi driver in testcase.yaml as well otherwise it will not be tested (your overlay is not used at all).

Also, **people have forgot to add test cases for their new drivers** for the following and reviewers forgot to mention that too :-)

- dcmi
- esp32_dvp
- smartdma

CharlesDias replied:
Hi, @ngphibang 

I can add this test case in another PR. I have to learn the test framework first. :-)

ngphibang commented on tests/drivers/build_all/video/stm32h7b3i_dk.overlay:
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2024, Charles Dias <charlesdias.cd@outlook.com>
```
2025

CharlesDias commented:
Rebase

CharlesDias commented:
Fix format code

CharlesDias commented:
@ngphibang and @josuah

The `bus_type` is initialized with the default value `VIDEO_BUS_TYPE_CSI2_DPHY` when `bus-type` property is not declared in the device tree. This way, I don't need to make changes on the video test folder for now.

I'll add the test case for DCMI driver in another PR. I have to study the test framework first! :)

josuah reviewed (approved):
Having a default bus type set to MIPI CSI allows to preserve backward compatibility, so no migration guide required I think.

Thank you for the few review/modification cycles! LGTM.

CharlesDias commented:
Hi @ngphibang and @josuah 

Commit message updated. Could you please remove the DNM flag? Thanks!

JarmouniA reviewed (approved):
Fantastic work, thanks @CharlesDias, let's get this in before the feature freeze.
